### ISSUE TITLE
NAN value appears during training

### ISSUE THREAD

**Issue by SanjuSoni** [lululxvi/deepxde#65]
dear @lululxvi 
I am trying to solve a non-dimensionalized coupled equation of order 1,

but while training the model I am getting several NAN values.

Kindly suggest how to fix it.

coupled eq, domain, BC, code,

 and error is as given below:

Error Message:

 ![error_p_non_dim](https://user-images.githubusercontent.com/61820415/84575948-af0db200-adce-11ea-8f69-0dcf31984855.JPG)

Code is :

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function


import sys
sys.path.insert( 0, "D:/PhD_Box/Code/Project_1/Stoke_2_DeepXDE/deepxde-master/deepxde-master")

import numpy as np

import deepxde as dde

import tensorflow as tf


def main():

    def pde( x, y ):

        Ra = 1e5
        Pr = 0.71
        pi = 3.14

        u, v, p, T = y[:, 0:1], y[:, 1:2 ], y[:, 2:3], y[:, 3:4]

        du_x = tf.gradients( u , x )[0]
        dv_x = tf.gradients( v , x )[0]
        dp_x = tf.gradients( p , x )[0]
        dT_x = tf.gradients( T , x )[0]


        dp_x, dp_y = dp_x[:, 0:1 ], dp_x[:, 1:2 ]

        du_x, du_y  = du_x[:, 0:1], du_x[:, 1:2 ]

        dv_x, dv_y = dv_x[:, 0:1], dv_x[ :, 1:2 ]

        dT_x, dT_y = dT_x[:, 0:1 ], dT_x[:, 1:2 ]


        du_xx = tf.gradients( du_x, x )[0][:, 0:1 ]
        du_yy = tf.gradients( du_y, x )[0][:, 1:2 ]

        dv_xx = tf.gradients( dv_x, x)[0][:, 0:1 ]
        dv_yy = tf.gradients( dv_y, x)[0][:, 1:2 ]

        dT_xx = tf.gradients( dT_x, x )[0][:, 0:1 ]
        dT_yy = tf.gradients( dT_y, x )[0][:, 1:2 ]



        return [
                du_x + du_y ,
                2*u*du_x + v*du_y+ dp_x - Pr*(du_xx + du_yy),
                du_x*v + u*dv_x + 2*v*dv_y + dp_y - Pr*(dv_xx + dv_yy )+ Ra*Pr*T,
                du_x*T +u*dT_x +dv_y-T + v*dT_y - dT_xx - dT_yy
                ]



    # function definition

    def f_zero(x):

        return np.zeros((len(x),1))

    def f_temp(x):

        pi = 3.14
        return  (1 - np.cos(2*pi*x[:,0:1]))/2


    rect_geom = dde.geometry.Rectangle([0,0], [ 12, 5 ])



    # boundary definition of a unit square domain

    def boundary_left( x, on_boundary ):

        return on_boundary  and np.isclose(x[0], 0 )

    def boundary_right( x, on_boundary ):

        return on_boundary  and np.isclose(x[0], 1 )

    def boundary_upper( x, on_boundary ):

        return on_boundary and np.isclose(x[1], 1 )

    def boundary_lower( x, on_boundary ):

        return on_boundary  and np.isclose(x[1], 0 )


    # defining BC for u, v, and p (component = 0, 1, 2 resp assign)

    bc_left_u = dde.DirichletBC( rect_geom, f_zero, boundary_left, component = 0 )
    bc_left_v = dde.DirichletBC( rect_geom, f_zero , boundary_left, component = 1 )
    bc_left_T = dde.DirichletBC( rect_geom, f_zero , boundary_left, component  = 3)



    bc_right_u = dde.DirichletBC( rect_geom, f_zero, boundary_right, component = 0 )
    bc_right_v = dde.DirichletBC( rect_geom, f_zero , boundary_right, component = 1 )
    bc_right_T = dde.DirichletBC( rect_geom, f_zero , boundary_right, component  = 3)


    bc_lower_u = dde.DirichletBC( rect_geom, f_zero, boundary_lower, component = 0 )
    bc_lower_v = dde.DirichletBC( rect_geom, f_zero, boundary_lower, component = 1 )
    bc_lower_T = dde.DirichletBC( rect_geom, f_temp, boundary_lower, component = 3 )

    bc_upper_u = dde.DirichletBC( rect_geom, f_zero, boundary_upper, component = 0 )
    bc_upper_v = dde.DirichletBC( rect_geom, f_zero, boundary_upper, component = 1 )
    bc_upper_T = dde.DirichletBC(rect_geom, f_zero, boundary_upper, component = 3)


    #bc_lwr_p = dde.DirichletBC(geom, lambda x: 0.00001 * np.ones((len(x), 1)), boundary_upr, component = 2)

    bc = [
             bc_left_u, bc_left_v, bc_left_T,
             bc_right_u, bc_right_v, bc_right_T,
             bc_lower_u, bc_lower_v, bc_lower_T,
             bc_upper_u, bc_upper_v, bc_upper_T,
            ]


    data = dde.data.PDE(
                     rect_geom,
                     pde,
                     bc,
                     num_domain = 1000,
                     num_boundary = 500,
                     num_test = 2000
                   )



    layer_size = [2] + [50] * 2 + [4]
    activation = "tanh"
    initializer = "Glorot uniform"

    net = dde.maps.FNN(layer_size, activation, initializer)

    model = dde.Model(data, net)

    model.compile( "adam", lr = 0.001 )

    losshistory, train_state = model.train(epochs = 5000)

    dde.saveplot(losshistory, train_state, issave=True, isplot=True)


if __name__ == "__main__":
    main()

```

======================================
Coupled equation and BC are:

-----------------------------------------------------------------

![cont_1](https://user-images.githubusercontent.com/61820415/84575985-f2682080-adce-11ea-8821-bbe3372dc093.JPG)
![cont_2](https://user-images.githubusercontent.com/61820415/84575990-f6943e00-adce-11ea-9628-155331851778.JPG)

- Kindly suggest me for fixing this issue.

**Comment by lululxvi** [comment]
Should it be
```python
def boundary_upper( x, on_boundary ):
    return on_boundary and np.isclose(x[1], 5)
```

**Comment by SanjuSoni** [comment]
Thanks @lululxvi 

Sorry for my mistake.

Actually, I had defined wrong geometry which should be:

``` rect_geom = dde.geometry.Rectangle([0,0], [ 1, 1]) ```

- Now, It's working fine except one place of loss value.


- During training steps, it has a huge value for boundary loss.

**The third column in train and test loss:**




![training_data](https://user-images.githubusercontent.com/61820415/84584990-3fc4ac00-ae28-11ea-9887-ad2376d99902.JPG)

![train loss](https://user-images.githubusercontent.com/61820415/84585035-e90ba200-ae28-11ea-93b2-2078d132185a.png)


 - Is it ok to have such a huge loss for this model?

Kindly, suggest me for that.

**Comment by lululxvi** [comment]
Maybe not. You need to check whether the solution is correct. If not, you can try to give a small weight for this loss via `loss_weights` in `Model.compile()`.

**Comment by SanjuSoni** [comment]
Thanks @lululxvi 

Is this the right way to deal with weight value in the model?

`Model.compile(optimizer, lr= 0.001,   loss_weights = 0.012)`


- what value can be suitable (approx). for loss_weights

**Comment by lululxvi** [comment]
See #15 #41 for ` loss_weights`, and also "Q: I failed to train the network, e.g., the training loss is large." in the FAQ.

**Comment by chenyv118** [comment]
> Maybe not. You need to check whether the solution is correct. If not, you can try to give a small weight for this loss via `loss_weights` in `Model.compile()`.

I try to adjust loss_weights in my model, but even if I set it to 0, nan will appear, and the first 15 of the 17 state variables are all nan
